TST: tests for maybe_promote (precursor to #23982)

h-vetinari opened the pull request:
First step towards #23833, resp. precursor to #23982.

TL;DR: `maybe_promote` is quite broken. #23982 tries to come up with tests that it should pass, and #25425 tries to fix the implementation. However, #23982 is quite big, so @jreback asked for a smaller version that (mostly) just tests existing behaviour.

h-vetinari committed:
TST: tests for maybe_promote (precursor to #23982)

pep8speaks commented:
Hello @h-vetinari! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/dtypes/cast/test_promote.py`](https://github.com/pandas-dev/pandas/blob/7801d0f1128ed8c4e8f2e076260ac6265a92e6ab/pandas/tests/dtypes/cast/test_promote.py):

> [Line 339:1](https://github.com/pandas-dev/pandas/blob/7801d0f1128ed8c4e8f2e076260ac6265a92e6ab/pandas/tests/dtypes/cast/test_promote.py#L339): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 342:1](https://github.com/pandas-dev/pandas/blob/7801d0f1128ed8c4e8f2e076260ac6265a92e6ab/pandas/tests/dtypes/cast/test_promote.py#L342): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 344:1](https://github.com/pandas-dev/pandas/blob/7801d0f1128ed8c4e8f2e076260ac6265a92e6ab/pandas/tests/dtypes/cast/test_promote.py#L344): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 349:1](https://github.com/pandas-dev/pandas/blob/7801d0f1128ed8c4e8f2e076260ac6265a92e6ab/pandas/tests/dtypes/cast/test_promote.py#L349): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 439:80](https://github.com/pandas-dev/pandas/blob/7801d0f1128ed8c4e8f2e076260ac6265a92e6ab/pandas/tests/dtypes/cast/test_promote.py#L439): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

jreback commented on pandas/tests/dtypes/cast/test_promote.py:
```diff
@@ -0,0 +1,624 @@
+
+    # try/except as numpy dtypes (i.e. if result_dtype is np.object_) do not
+    # know some expected dtypes like DatetimeTZDtype, and hence raise TypeError
+    try:
```
I find this very error prone an would rather make this explicit

h-vetinari replied:
I don't understand what you mean here. PandasDtypes (which will not be known by a numpy dtype) can come in on either side of the comparison. If you have a way to write this better, let me know.

gfyoung replied:
I think you need to check the type of `result_dtype` and `expected_dtype`.  That way, you know which way to compare.

jreback commented on pandas/tests/dtypes/cast/test_promote.py:
```diff
@@ -0,0 +1,624 @@
+        assert result_dtype == expected_dtype
+    except TypeError:
+        assert expected_dtype == result_dtype
+
```
handle null values in a separate clause to make this less compound

h-vetinari replied:
OK

jreback commented on pandas/tests/dtypes/cast/test_promote.py:
```diff
@@ -0,0 +1,624 @@
+@pytest.mark.parametrize('boxed, box_dtype', [
+    (True, None),    # fill_value wrapped in array with auto-dtype
+    (True, object),  # fill_value wrapped in array with object dtype
+    (False, None)    # fill_value directly
```
what does box_dtype of None even mean?

h-vetinari replied:
I'm trying to explain just that with the comment next to it.

h-vetinari replied:
Side note here: None takes no extra meaning, but is passed directly to the constructor of `np.array(some_value, dtype=None)`, which corresponds to the default behaviour of numpy.

gfyoung replied:
Incorporate this side-note as a comment.  That will help to explain the parameterized values better.

jreback reviewed (changes requested):
just started glancing at this. I find the box & box_dtype parametrization args very confusing. This is testing a scalar and array, but when / where can this actually be an object dtype of a scalar? I would like to keep the scope of this way way down. This means don't over parmaterize.

h-vetinari reviewed (commented):
> @jreback: just started glancing at this. I find the box & box_dtype parametrization args very confusing. This is testing a scalar and array, but when / where can this actually be an object dtype of a scalar? I would like to keep the scope of this way way down. This means don't over parmaterize.

Thanks for starting to take a look at this. I intentionally kept these tests together, as the actual tests are the same in 95% of the cases, and it would lead to huge amount of duplication to separate testing the scalar case vs. the array case.

I've tried to annotate at every step what `boxed` and `box_dtype` mean, if we can find a better way to write this I'm all ears, but I don't think breaking up the tests is a good approach.

jreback commented:
though a good idea, closing as stale

h-vetinari commented:
> @jreback: though a good idea, closing as stale

I've worked an inordinate amount of time on #25425, and although I prepared two subsets of it (#23982 and this one), it never got any review. I understand the realities of how scarce reviewing resources are, but it's still disheartening after putting in so much work.

Long story short, all those three PRs (#23982,#25425, #25637) should be reopened, please. I'm ready to keep working on them.

gfyoung commented:
@h-vetinari : I'll re-open this one first.  I agree with @jreback that the scope of what you're trying to do is massive.  Thus, it would be great if you focused on this PR alone.

We can re-open other ones as the work progresses.

gfyoung commented:
> it would lead to huge amount of duplication to separate testing the scalar case vs. the array case.

In a separate commit, separate these out as requested.  The best way to make the point here is to show the code side-by-side.  Are you sure you can't abstract out setup methods for some of the functionality that would be de-duplicated by any chance?

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into tst_maybe_promote_precursor

h-vetinari committed:
Fix tabs vs. spaces

h-vetinari commented:
@gfyoung 
Thanks for reopening.

> In a separate commit, separate these out as requested.

I can do this, but essentially *every* test would be duplicated. I'm testing the same logic in both cases, and actually more than that: *that* the same promotion logic applies in both scalar and array case.

If I start splitting this, it would make most sense to have one test for scalars and one for arrays, but from that point on, it would be very easy for those tests to diverge. Does this sound like a reasonable trade-off to you? While trivial, it will be a fair bit of work to actually duplicate these, so I'd rather be sure you understand the consequences before-hand.

>  Are you sure you can't abstract out setup methods for some of the functionality that would be de-duplicated by any chance?

That's exactly what I was aiming for with `_check_promote`. All the "differences" between the array and scalar case are taken care of by the wrapper, so that each test *only* has to specify the promotion logic that's being tested.

gfyoung commented:
> If I start splitting this, it would make most sense to have one test for scalars and one for arrays, but from that point on

Why don't you try duplicating for one test only.  Then we can evaluate before and after on that one test and see whether it makes sense for you to continue forward with the duplication.

> That's exactly what I was aiming for with _check_promote.

Hmm...I see.  It sounds though (from the comments IIUC) that that might have impacted readability because there were so many different cases to check.  Is it possible to maybe have two or three different check promote functions that each handle a subset of cases?

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into tst_maybe_promote_precursor

h-vetinari committed:
fix conftest

h-vetinari committed:
more conftest merge conflict artefact fixes

h-vetinari committed:
lint

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into tst_maybe_promote_precursor

h-vetinari committed:
Review (jreback)

h-vetinari committed:
Add docstring to _check_promote

h-vetinari commented:
@gfyoung 

Finally got around to coming back to this:

> > That's exactly what I was aiming for with _check_promote.

> Hmm...I see. It sounds though (from the comments IIUC) that that might have impacted readability because there were so many different cases to check. Is it possible to maybe have two or three different check promote functions that each handle a subset of cases?

I added a docstring and some comments to make it clearer. The function essentially only contains one branch depending on whether `fill_value` should be passed as scalar or as an array. And that's exactly the part I want to abstract away from the tests further on.

All of those tests have a very simple format:
```python
def test_maybe_promote_XXX_with_YYY(fixtures..., boxed, box_dtype):
    dtype = some_dtype(XXX)
    fill_dtype = some_dtype(YYY)

    if some_condition:
        pytest.xfail('reason')

    # define fill_value (usually implicitly through fill_dtype)
    fill_value = some_scalar(fill_dtype)

    # define expected values
    expected_dtype = some_dtype
    exp_val_for_scalar = some_scalar_value
    exp_val_for_array = some_missing_value_marker

    _check_promote(dtype, fill_value, boxed, box_dtype, expected_dtype,
                   exp_val_for_scalar, exp_val_for_array)
```

This about as simple as I can imagine it to be (while unifying the treatment of scalar and array case). Short of a new idea (and the docstring), I don't know how to make this clearer.

(speaking of new ideas: I incorporated your typecheck instead of the try/except for `PandasExtensionDtype`)

gfyoung commented on pandas/conftest.py:
```diff
@@ -409,16 +421,16 @@ def tz_aware_fixture(request):
-DATETIME_DTYPES = ['datetime64[ns]', 'M8[ns]']
-TIMEDELTA_DTYPES = ['timedelta64[ns]', 'm8[ns]']
+DATETIME64_DTYPES = ['datetime64[ns]', 'M8[ns]']
+TIMEDELTA64_DTYPES = ['timedelta64[ns]', 'm8[ns]']
```
Rationale for renames?

Is this necessary given how massive the diff is already?

h-vetinari replied:
I can split off the conftest-related things, of course.

Regarding your specific point: The rename is really important IMO, because the `DATETIME_DTYPES` (as-is) have nothing to do with `datetime.dateime` or other such things.

However, when reading a test that is being fed by a `datetime_dtype` fixture, one could easily think that this tests `datetime.datetime`.

The current state in conftest specifically contains only `DATETIME64_DTYPES`, and I believe this distinction is used in many parts of the code base as well.

gfyoung replied:
Got it.  I'm thinking that we should separate this out.  The more we can condense this PR, the better.

But hold off for the moment (just one more question re: some of other things you've changed for `conftest.py`.

h-vetinari replied:
I can change the `DATETIME_DTYPES` -> `DATETIME64_DTYPES` in a precursor of course, but the actual change that precipitates the necessity is the introduction of `datetime64_dtype`, which wouldn't be used before this PR. Would you want me to introduce the fixtures with the conftest-precursor, or in this PR?

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into tst_maybe_promote_precursor

h-vetinari committed:
doc improvements

h-vetinari committed:
fix c&p artefact

h-vetinari committed:
Introduce parametrized fixture for box-variants; override where required

h-vetinari commented:
> @gfyoung: [Incorporate this side-note as a comment. That will help to explain the parameterized values better.](https://github.com/pandas-dev/pandas/pull/25637/files#r289039845)

I started adding more explanations, but since that would have been repeated for all tests, I decided it's cleaner to have a parametrized fixture that only needs to be explained once (and is overridden where necessary due to xfails or special cases).

I think this is much cleaner now. :)

h-vetinari committed:
improve docstring for box-fixture

h-vetinari committed:
fix comments

gfyoung commented on pandas/conftest.py:
```diff
@@ -376,28 +376,40 @@ def unique_nulls_fixture(request):
+                'dateutil.tz.tzutz()', 'dateutil.tz.tzlocal()',
+                'pytz.FixedOffset(300)', 'pytz.FixedOffset(0)',
+                'pytz.FixedOffset(-300)', 'datetime.timezone.utc',
+                'datetime.timezone.+1', 'datetime.timezone.-1.named']
```
Remind me again why `TIMEZONE_IDS` are needed?

h-vetinari replied:
@gfyoung 
Fair point (as with the other conftest-related comments), but it was needed for me because debugging tz-tests is very tedious without knowing for which timezone the tests failed. 

Related discussion: https://github.com/pandas-dev/pandas/pull/25425#discussion_r263968845

gfyoung replied:
Got it.  Can you open a PR that breaks off the following changes:

* `TIMEZONE_IDS`
* `DATETIME_DTYPES` rename
* `TIMEDELTA_DTYPES` rename

It will be a small PR, but I think it's a reasonable change to be made independently of this PR and shrink it by a little bit.

h-vetinari reviewed (commented):
Happy to split off whatever from conftest, just need to tell me how you want it exactly. :)

h-vetinari committed:
fix fixture docstrings

h-vetinari committed:
docstring updates

h-vetinari committed:
add ids for box-fixture

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into tst_maybe_promote_precursor

h-vetinari commented:
@gfyoung 
This is now slightly reduced in size after #26596. Also, could you please restart the azure CI?

h-vetinari committed:
lint

gfyoung commented:
@jreback : Could you have another look at this

jreback commented:
at some point

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into tst_maybe_promote_precursor

h-vetinari committed:
break out function to make _check_promote even more branchless

jreback commented:
so why are you haveing *every* int type against a timedelta64 for example; this is just way overkill. These should all react the same, but having this expanse of tests is just too confusing. Please limit this.

```
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint8-timedelta64[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint8-timedelta64[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint8-m8[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint8-m8[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint16-timedelta64[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint16-timedelta64[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint16-m8[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint16-m8[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint32-timedelta64[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint32-timedelta64[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint32-m8[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint32-m8[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint64-timedelta64[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint64-timedelta64[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint64-m8[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[uint64-m8[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[int-timedelta64[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[int-timedelta64[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[int-m8[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[int-m8[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[int8-timedelta64[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[int8-timedelta64[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[int8-m8[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[int8-m8[ns]-box2-np.timedelta64]
  reason: does not upcast correctly
XFAIL pandas/tests/dtypes/cast/test_promote.py::test_maybe_promote_any_with_timedelta64[int16-timedelta64[ns]-box0-np.timedelta64]
  reason: does not upcast correctly
```

h-vetinari commented:
> @jreback: so why are you haveing every int type against a timedelta64 for example; this is just way overkill. These should all react the same, but having this expanse of tests is just too confusing. Please limit this.

The simple answer is: fixtures - they do exactly what I wanted here: exhaustively (as possible) test every type against (inputs of) every other type. I get the argument that some tests can be overkill, but I can't see why it would be confusing...?

Are you opposed specifically to the number of xfails, I'm guessing? Otherwise, once one starts introducing special cases in this module it becomes less understandable, more fragile, and more complex.

jreback commented:
> The simple answer is: fixtures - they do exactly what I wanted here: exhaustively (as possible) test every type against (inputs of) every other type. I get the argument that some tests can be overkill, but I can't see why it would be confusing...?

@h-vetinari this add quite a bit of duplicative testing for many types. Please use more selective fixtures. If its obvious that something is doing the same exact cast vs int, int8, int16.....then a single int is enough for a fixture with a nice comment.

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into tst_maybe_promote_precursor

h-vetinari committed:
Refactor with reduced any_numpy_dtype fixture (review jreback)

h-vetinari commented:
@jreback 
What do you think of the reduced fixture I introduced following your review?

jreback commented:
ok @h-vetinari whats 3000+ new tests, lol.

merge master once again and ping on green.

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into tst_maybe_promote_precursor

h-vetinari commented:
> merge master once again and ping on green.

@jreback, this is green. :)

jreback commented:
k thanks @h-vetinari

h-vetinari commented:
@jreback @gfyoung 
Thanks for reviewing this big hunk. :)

> We can re-open other ones as the work progresses.

Can you please reopen #25425?

I'd also suggest that we reopen #23982 anyway - I don't think it has a chance of getting merged (because it basically just adds a bunch of failing tests), but once rebased it would provide a good insight for the _value_ of the refactor #25425 - as in: all the things that are currently broken and would be fixed by that PR.

h-vetinari commented:
@jreback @gfyoung 

> @h-vetinari: Can you please reopen #25425?
